cmd/xl-v1.go

```diff
@@ -195,45 +200,225 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
 	return &Metrics{}, NotImplemented{}
 }
 
-// CrawlAndGetDataUsage picks three random disks to crawl and get data usage
-func (xl xlObjects) CrawlAndGetDataUsage(ctx context.Context, endCh <-chan struct{}) DataUsageInfo {
-	var randomDisks []StorageAPI
+// CrawlAndGetDataUsage will start crawling buckets and send updated totals as they are traversed.
+// Updates are sent on a regular basis and the caller *must* consume them.
+func (xl xlObjects) CrawlAndGetDataUsage(ctx context.Context, updates chan<- DataUsageInfo) error {
+	cache := make(chan dataUsageCache, 1)
+	defer close(cache)
+	go func() {
+		for update := range cache {
+			updates <- update.dui(update.Info.Name)
+		}
+	}()
+	return xl.crawlAndGetDataUsage(ctx, cache)
+}
+
+// CrawlAndGetDataUsage will start crawling buckets and send updated totals as they are traversed.
+// Updates are sent on a regular basis and the caller *must* consume them.
+func (xl xlObjects) crawlAndGetDataUsage(ctx context.Context, updates chan<- dataUsageCache) error {
+	var disks []StorageAPI
+	buckets, err := xl.ListBuckets(ctx)
+	if err != nil {
+		return err
+	}
 	for _, d := range xl.getLoadBalancedDisks() {
 		if d == nil || !d.IsOnline() {
 			continue
 		}
-		randomDisks = append(randomDisks, d)
-		if len(randomDisks) >= 3 {
-			break
+		disks = append(disks, d)
+	}
+	if len(disks) == 0 || len(buckets) == 0 {
+		return nil
+	}
+
+	// Load bucket totals
+	oldCache := dataUsageCache{}
+	var buf bytes.Buffer
+	err = xl.GetObject(ctx, minioMetaBackgroundOpsBucket, dataUsageCacheName, 0, -1, &buf, "", ObjectOptions{})
+	if err != nil {
+		if !isErrObjectNotFound(err) {
+			return toObjectErr(err, minioMetaBackgroundOpsBucket, dataUsageCacheName)
+		}
+	} else {
+		err := oldCache.deserialize(buf.Bytes())
+		if err != nil {
+			logger.LogIf(ctx, err)
+			oldCache = dataUsageCache{}
+		}
+		//fmt.Println("old cache deserialized:", oldCache.StringAll())
```

Comment:
perhaps you wanted to add `if dataUsageDebug` ? - not sure if we should keep these `fmt.Println` in the code if there is no way to enable them?